pymongo/client_session.py

```diff
@@ -368,6 +379,128 @@ def _inherit_option(self, name, val):
+        should be mindful of this possiblity when writing a ``callback`` that
+        modifies application state or has any other side-effects.
+        Note that even when the ``callback`` is invoked multiple times,
+        ``with_transaction`` ensures that the transaction will be committed
+        at-most-once on the server.
+
+        The ``callback`` should not attempt to start new transactions, but
+        should simply run operations meant to be contained within a
+        transaction. The ``callback`` should also not commit the transaction;
+        this is handled automatically by ``with_transaction``. If the
+        ``callback`` does commit or abort the transaction without error,
+        however, ``with_transaction`` will return without taking further
+        action.
+
+        When ``callback`` raises an exception, ``with_transaction``
+        automatically aborts the current transaction. When ``callback`` or
+        :meth:`~ClientSession.commit_transaction` raises an exception that
+        includes the ``"TransientTransactionError"`` error label,
+        ``with_transaction`` starts a new transaction and re-executes
+        the ``callback``.
+
+        When :meth:`~ClientSession.commit_transaction` raises an exception with
+        the ``"UnknownTransactionCommitResult"`` error label,
+        ``with_transaction`` retries the commit until the result of the
+        transaction is known.
+
+        This method will cease retrying after 120 seconds has elapsed. This
+        timeout is not configurable and any exception raised by the
+        ``callback`` or by :meth:`ClientSession.commit_transaction` after the
+        timeout is reached will be re-raised. Applications that desire a
+        different timeout duration should not use this method.
+
+        :Parameters:
+          - `callback`: The callable ``callback`` to run inside a transaction.
+            The callable must accept a single argument, this session. Note,
+            under certain error conditions the callback may be run multiple
+            times.
+          - `read_concern` (optional): The
+            :class:`~pymongo.read_concern.ReadConcern` to use for this
+            transaction.
+          - `write_concern` (optional): The
+            :class:`~pymongo.write_concern.WriteConcern` to use for this
+            transaction.
+          - `read_preference` (optional): The read preference to use for this
+            transaction. If ``None`` (the default) the :attr:`read_preference`
+            of this :class:`Database` is used. See
+            :mod:`~pymongo.read_preferences` for options.
+
+        :Returns:
+          The return value of the ``callback``.
+
+        .. versionadded:: 3.9
+        """
+        start_time = monotonic.time()
+        while True:
+            self.start_transaction(
+                read_concern, write_concern, read_preference)
+            try:
+                ret = callback(self)
```

Comment:
@ajdavis does this implementation work for motor? Will motor have to rewrite `with_transaction` no matter what because the `callback` will be an async method?

Reply:
I think Motor can reuse this. Motor should accept a coroutine as a callback, wrap that in another callback which schedules the coroutine on the event loop and waits (blocking this thread) until the coroutine finishes, then passes back its return value or reraises its exception, unblocking this thread.

Once it's wrapped the coroutine in the wrapper callback, Motor will do its usual thing: schedule PyMongo's with_transaction() on the Motor thread pool and return a Future that's resolved when with_transaction() finishes. I can help whoever does this implementation.

Reply:
Perfect, that makes sense to me.